### PR TITLE
master/steps: add a CMake step

### PR DESCRIPTION
**Pull request by sa2ajj**
This step provides a convenience interface for `cmake` command.

### CODE REVIEW

**Comment by tardyp** [comment]
looks good so far!

File: master/docs/manual/cfg-buildsteps.rst [hunk 1]
```diff
@@ -1240,8 +1267,65 @@ The arguments are identical to :bb:step:`ShellCommand`.
...
+
+In addition to the parameters :bb:step:`ShellCommand` supports, this step accepts the following parameters:
+
+``path``
+    Either a path to a source directory to (re-)generate a build system for it in the current working directory.
+    Or an existing build directory to re-generate its build system.
+
+``generator``
+    A build system generator.
+    See `cmake-generators(7) <https://cmake.org/cmake/help/v3.5/manual/cmake-generators.7.html>`_ for available options.
```
Comment (Stibbons) [inline master/docs/manual/cfg-buildsteps.rst]: Why would you run the generator in a CI ?

Comment (sa2ajj) [inline master/docs/manual/cfg-buildsteps.rst]: I am not sure I understood the question: in CMake a generator is what produces the actual build system.  In my day job, it's Ninja.

Could you please clarify what you mean?

Comment (Stibbons) [inline master/docs/manual/cfg-buildsteps.rst]: My mistake. I remember the way of using cmake and wasn't traditional...

I mean, there might be several strategy around cmake:
- indeed, you might want to run the generator in the ci to product the build system files (Makefile, Ninja,...). In the CI, you need to run the generator and then the build system. => for that, don't you want to add a way to run the build system automatically, or you want user to add another "build" step that would run the ninja command?
- in my case, we generated the makefiles once, committed them, and made heavy modifications on it, so we didn't run the generator in the CI... but, yes, it's not the canonic way of working with cmake !

**Comment by sa2ajj** [comment]
Looks like I need to spend a bit more time on figuring out some class dependencies :smile:

File: master/docs/manual/cfg-buildsteps.rst (hunk 1 above)
Comment (sa2ajj) [inline master/docs/manual/cfg-buildsteps.rst]: Our  case is that we *require* users to first run CMake and then run ninja:
```
$ git clone remote:repo
$ cd repo
$ mkdir build
$ cd build
$ cmake -G Ninja .. <other parameters>
$ ninja
$ ninja unittest.run
```

We have 6 different target environments (4 hardware + 2 desktop) + a number of options that control, for example, static analysis target generation in `build.ninja` file.

All this somehow makes us feel that adding all those pre-configured environments does not make sense.

**Comment by sa2ajj** [comment]
Any suggestions on negative test cases?

File: master/buildbot/test/unit/test_steps_cmake.py [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+        self.expectOutcome(result=SUCCESS)
+        return self.runStep()
+
+    def test_definitions(self):
+        definition = {
+            'a': 'b'
+        }
+        self.setupStep(CMake(definitions=definition))
+        self.expect_command('-D%s=%s' % definition.items()[0])
+
```
Comment (tardyp) [inline master/buildbot/test/unit/test_steps_cmake.py]: you are supposed to run your step!

return self.runStep()

**Comment by tardyp** [comment]
not sure what you mean by negative test cases.

File: master/buildbot/test/unit/test_steps_cmake.py (hunk 2 above)
Comment (sa2ajj) [inline master/buildbot/test/unit/test_steps_cmake.py]: I should rename the function name since it's 'expect **and** run`

**Comment by sa2ajj** [comment]
any suggestions of what **wrong** situations test cases should try?

**Comment by tardyp** [comment]
LGTM. Testing the config parameters is good enough for me.

You got 100% hit on the patch

**Comment by rutsky** [comment]
@sa2ajj I see that you reformatted some RST files, do you use some tools do this or do you reformat them manually?

**Comment by sa2ajj** [comment]
@rutsky I do it manuall :)  I use ReST quite a lot and these kinds of problems stand out for me and my fingers automatically fix them :)

File: master/buildbot/steps/cmake.py [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+            for item in iteritems(self.definitions):
+                command.append('-D%s=%s' % item)
+
+        if self.options is not None:
+            command.extend(self.options)
+
+        cmd = yield self.makeRemoteShellCommand(
+            command=command,
+            collectStdout=True,
+            collectStderr=True
```
Comment (djmitche) [inline master/buildbot/steps/cmake.py]: collect* mean that these outputs will be stored in memory, and I imagine they'll be quite large -- but they appear not to be used.  Is this necesary?

**Comment by djmitche** [comment]
This looks good to me, with that one question answered.  I didn't look to hard at most of the .rst except the bit about the new step.

File: master/buildbot/steps/cmake.py (hunk 3 above)
Comment (sa2ajj) [inline master/buildbot/steps/cmake.py]: Oh, in this case it's not really necessary.

I need to check the docs about those options and this note if it's not mentioned.

Comment (sa2ajj) [inline master/buildbot/steps/cmake.py]: Ok, it's somewhat mentioned :/

I updated this branch to not use these parameters.

**Pull request merged**